src/midgard/polyline2.cc

```diff
@@ -42,17 +44,209 @@ typename coord_t::value_type Polyline2<coord_t>::Length(const container_t& pts)
+          printf("Avoiding self-intersection:\n");
+          printf("sidx: %lu, s: %.6f, %.6f\n", sidx, start.lat(), start.lng());
+          printf("eidx: %lu, e: %.6f, %.6f\n", eidx, end.lat(), end.lng());
+          printf("cidx: %lu, c: %.6f, %.6f\n", cidx, c.lat(), c.lng());
+          printf("pidx: %lu, p: %.6f, %.6f\n", point_idx, p.lat(), p.lng());
+          triangle_contains(start, c, end, p);
+#endif
+          break;
+        }
+      }
+
+      // the moment we realize we cannot simplify we can stop
+      if (!can_simplify) {
+        break;
+      }
+    }
+
+    if (can_simplify) {
+      // Simplify the polyline by removing all points between sidx and eidx
+      // from the point-tile-index (but don't remove sidx or eidx).
+      point_tile_index.remove_points(sidx + 1, eidx);
+    } else {
+      // Simplifying this polyline would result in a self-intersection, so
+      // we cannot. Force recursion around hfidx.
+      dmax = epsilon_sq;
+    }
+  }
+
+  // if (dmax >= epsilon_sq) there are some high frequency details between start
+  // and end so we need to look for flatter sections between them.
+  if (dmax >= epsilon_sq) {
+    // we recurse from right to left for two reasons:
+    // 1. we want to preserve iterator validity in the vector version
+    // 2. its the only way to preserve the indices in the keep set
+    if (eidx - hfidx > 1)
+      peucker_avoid_self_intersections(point_tile_index, epsilon_sq, exclusions, hfidx, eidx);
+    if (hfidx - sidx > 1)
+      peucker_avoid_self_intersections(point_tile_index, epsilon_sq, exclusions, sidx, hfidx);
+  }
+}
+
+template <class coord_t, class container_t>
+void DouglastPeuckerAvoidSelfIntersection(container_t& polyline,
+                                          typename coord_t::value_type epsilon_m,
+                                          const std::unordered_set<size_t>& exclusions) {
+  // Create a tile-space-index which allows us to perform a Douglas-Peucker style line
+  // simplification that avoids creating self-intersections within the polyline/polygon.
+  const PointLL& first_point = *polyline.begin();
+  double meters_per_deg = DistanceApproximator<PointLL>::MetersPerLngDegree(first_point.lat());
+  double epsilon_deg = epsilon_m / meters_per_deg;
+  PointTileIndex point_tile_index(epsilon_deg);
+  point_tile_index.tile<container_t>(polyline);
+
+  peucker_avoid_self_intersections(point_tile_index, epsilon_m * epsilon_m, exclusions, 0,
+                                   polyline.size() - 1);
+
+  // copy the simplified 'points' into 'polyline'
+  polyline.clear();
+  for (const auto& pt : point_tile_index.points) {
+    if (pt != PointTileIndex::deleted_point) {
+      polyline.push_back(pt);
+    }
+  }
+}
+
+template <class coord_t, class container_t>
+void DouglasPeucker(container_t& polyline,
+                    typename coord_t::value_type epsilon,
+                    const std::unordered_set<size_t>& indices) {
```

Comment:
what do you think about unifying parameters names(e.g use `exclusions` here as well)

Reply:
Fair suggestion.  As to my thinking... this was the original routine, so I wanted to leave it unmodified.  Honestly tho, `indicies` just isn't very descriptive, so I am happy to rename it to `exclusions`.  I'll also give consideration to other variables that might need better names.

Reply:
>  this was the original routine, so I wanted to leave it unmodified
Agreed, I prefer not to modify the code I don't touch. It's just the specificity of this case that you introduce similar functions. Of course, this is a "soft suggestion"(non-mandatory). I'm ok with both outcomes(unified/not unified)